### ISSUE TITLE
Adding duplicate versions of replaced datastreams

### ISSUE THREAD

**Issue by bryjbrown** [SFULibrary/islandora_datastream_crud#64]
Recently we used @bgilling's https://github.com/ulsdevteam/islandora_datastreams_io module to do batch replacements of MODS datastreams on our repository, and the Solr indexing backlog exploded. After investigating, it looks like for every replacement the process ended up appending 2-6 identical MODS records, each one being identical copies of the one supplied for replacement. https://github.com/ulsdevteam/islandora_datastreams_io is basically a web UI for islandora_datastreams_crud, so I decided to test islandora_datastreams_crud directly via drush and found that consistently for any given datastream I replace, it actually replaces it multiple times. 

Example for reproducing: On a new VM, create a new PDF cmodel object and check the manage datastreams menu. The MODS record and OBJ PDF should both have 1 under "VERSIONS". If you use islandora_datastream_crud's drush command to run a replacement, the replacements will be successful but the number under "VERSIONS" will jump up to 3 or more, instead of the expected number of 2 (one old version, one new version). 

Since Fedora is keeping copies of these files this becomes a storage issue, and when replacing MODS records this crushes Solr by tripling (or more) the amount of reindexing it thinks it needs to do.

**Comment by wgilling** [comment]
Bryan,

I can explain why this happens... in updating most datastreams, the mimetype, the label, and the underlying resource (a MODS file in this case) are EACH creating another instance of the MODS datastream.


I hope this answers at least the "weird behavior" part of your question... as to what to do to address this, some code could be written to dispose of any datastream version where the underlying resource has not changed (the versions related to changing label and mimetype values could safely be discarded).  This however would do nothing to prevent Solr from indexing that object datastream 3 times.

I believe that this might be issued as a bug in the tuque library -- I believe that some checking can be done within the innermost tuque code so that it does not set a label attribute on the datastream if it is the same as the current value... the same conditional could be wrapped around setting the mimetype.

**Comment by mjordan** [comment]
@bryjbrown I addressed this problem in #58 but perhaps I didn't do a good enough job of testing my fix. Are you running the latest version of the CRUD code?

**Comment by mjordan** [comment]
But I was unaware of the issue @bgilling describes.

**Comment by bryjbrown** [comment]
@mjordan I'm using a freshly downloaded copy for testing.

@bgilling This is also happening when I replaced the OBJ datastream on PDF cmodel, so I don't think its MODS only, even though the problem is certainly exacerbated when dealing with MODS datastreams. Is islandora_datastreams_io the thing creating the new MODS versions then, or tuque?

When I use /datastream/MODS/replace, I only get 1 additional MODS version though, so it can't be that ANY old MODS replacement is triggering multiple versions. The UI is doing things as expected, so there's got to be something thats different about the way crud/io is handling this programmatically.

**Comment by mjordan** [comment]
Datastream CRUD is via the tuque API. Apparently my fix in #58 didn't work.

I'll make some time to look at this in a few hours, can't at the moment.

**Comment by mjordan** [comment]
@bryjbrown I can't replicate the duplication. Here's my datastream list after pushing a new version of MODS using the command `drush islandora_datastream_crud_push_datastreams --user=admin --datastreams_source_directory=/tmp/crud64mods`:

![crud64](https://user-images.githubusercontent.com/403918/44003437-82afefe4-9e07-11e8-8248-d2aa2027806a.png)

Following up on @bgilling's comment, are you specifying a mimetype? When I specify one, e.g. `drush islandora_datastream_crud_push_datastreams --user=admin --datastreams_source_directory=/tmp/crud64mods --datastreams_mimetype=application/xml`, I get two new versions. Specifying the mimetype is not necessary if you don't want to alter the datastream's mimetype; if omitted, the existing mimetype is retained.

**Comment by bryjbrown** [comment]
@mjordan this helps a lot, with this info I was able to figure out the problem. `islandora_datastreams_io` is calling `islandora_datastreams_crud` to push and specifying not only the mimetype, but also the datastream label. Each option seems to make a new version of the datastream and commenting them out only increments the version count by 1, as expected, while retaining the original mimetype and label. I'm working on a PR for `islandora_datastreams_io` now that incorporates this fix.

Now that I understand the problem, I don't think this is a bug with `islandora_datastreams_crud` since I was specifying DSID mimetype and label when testing, not knowing they were optional. Closing this issue now.
